### ISSUE TITLE
`TestCase.assertEqual` has `equal_nan` default to `True`

### ISSUE THREAD

**Issue by eqy** [pytorch/pytorch#72194]
### 🐛 Describe the bug

While [`torch.testing.assert_close`](https://pytorch.org/docs/stable/testing.html#:~:text=torch.testing.assert_close%28actual%2C%20expected%2C%20%2A%2C%20allow_subclasses%3DTrue%2C%20rtol%3DNone%2C%20atol%3DNone%2C%20equal_nan%3DFalse%2C%20check_device%3DTrue%2C,%5Bsource%5D%20Asserts%20that%20actual%20and%20expected%20are%20close.), [`torch.allclose`](https://pytorch.org/docs/stable/generated/torch.allclose.html), and [`torch.isclose`](https://pytorch.org/docs/stable/generated/torch.isclose.html) have `equal_nan` defaulted to `False`, it is default `True` in [`TestCase.assertEqual`](https://github.com/pytorch/pytorch/blob/2d5296b0e7769b9bd7cda9607c1d14820e9103e4/torch/testing/_internal/common_utils.py#L2075). While likely extremely rare, this default setting has lead to unexpected behavior e.g., (#72157) in the form of bogus comparisons between two `NaN` tensors. Should this default be revisited and the cases where `NaN`s in testing results expected/tolerated explicitly marked?

CC @ptrblck @xwang233 @zasdfgbnm @ngimel 

### Versions

Current upstream

**Comment by mruberry** [comment]
Yes we would generally accept PRs making `assertEqual` stricter and more similar to `testing.assert_close`, but the test suite will have to be updated to pass with the stricter defaults, too.
